File: inheritance/person/person.cpp
/*.
"Student" (Öğrenci) ve "Teacher" (Öğretmen) sınıflarını "Person" sınıfından türetin.
"Student" sınıfı, öğrenci numarası için ek bir özelliğe sahip olmalıdır.
"Teacher" sınıfı, öğrettikleri ders için ek bir özelliğe sahip olmalıdır.
Her sınıf, bilgilerini görüntülemek için bir yönteme sahip olmalıdır.
*/
#include <iostream>
#include <string>
using std::cout;
using std::endl;
using std::string;

class Person
{
protected:
    string name;
    int age;

public:
    string getName()
    {
        return name;
    };
    int getAge()
    {
        return age;
    };
    void setName(string name)
    {
        this->name = name;
    }
    void setAge(int age)
    {
        this->age = age;
    }
    Person()
    {
        cout << "Person ctor created" << endl;
    }
    Person(string name, int age) : Person()
    {
        cout << "full parameter Person ctor created" << endl;
        this->name = name;
        this->age = age;
    }
    virtual void displayInfo()
    {
        cout << "İsim: " << name << ", Yaş: " << age << endl;
    }
};

class Student : public Person
{
private:
    string schoolNumber;

public:
    Student(string name, int age, string schoolNumber) : Person(name, age)
    {
        this->schoolNumber = schoolNumber;
    }
    void displayInfo() override
    {
        cout << "Ogrenci Bilgileri -> Isim: " << name << ", Yas: " << age << ", Ogrenci Numarasi: " << schoolNumber << endl;
    }
};
class Teacher : public Person
{
private:
    string subject;

public:
    Teacher(string name, int age, string subject) : Person(name, age)
    {
        this->subject = subject;
    }
    void displayInfo() override
    {
        cout << "Ogretmen Bilgileri -> Isim: " << name << ", Yas: " << age << ", Ders: " << subject << endl;
    }
};

int main()
{
    Student student = Student("Ahmet", 20, "1235");
    Teacher teacher = Teacher("Elif", 35, "Matematik");
    student.displayInfo();
    teacher.displayInfo();
    char c = getchar();
    return 0;
}

File: inheritance/person/person.py
class Person:
    def __init__(self,name,age):
        self.__name = name
        self.__age = age
        print("Person Nesnesi oluşturuldu")
    
    @property
    def name(self):
        return self.__name
    
    @name.setter
    def name(self,name):
        self.__name = name        

    @property
    def age(self):
        return self.__age
    
    @age.setter
    def age(self,age):
        self.__age = age
    
    def display_info(self):
        print(f"İsim: {self.name}, Yaş : {self.age}")

class Student(Person):
    def __init__(self,name,age,school_number):
        super().__init__(name,age)
        print("Student Nesnesi oluşturuldu")
        self.__school_number = school_number
    
    @property
    def school_number(self):
        return self.__school_number
    
    @school_number.setter
    def school_number(self,school_number):
        self.__school_number = school_number
    
    def display_info(self):
        print(f"Öğrenci Bilgileri -> İsim: {self.name}, Yaş: {self.age}, Öğrenci Numarası: {self.school_number}") 



class Teacher(Person):
    def __init__(self, name, age, subject):
        super().__init__(name, age)
        print("Teacher Nesnesi oluşturuldu")
        self.__subject = subject

    @property
    def subject(self):
        return self.__subject
    
    @subject.setter
    def subject(self,subject):
        self.__subject = subject

    def display_info(self):
        print(f"Öğretmen Bilgileri -> İsim: {self.name}, Yaş: {self.age}, Ders: {self.subject}")



if __name__ == "__main__":
    student = Student("Ahmet", 20, "1235")
    student.display_info()
    teacher = Teacher("Elif", 35, "Matematik")
    teacher.display_info()